src/cvrp/ecvrp.py

```diff
@@ -80,58 +90,415 @@ def get_roads(self) -> tuple[tuple[int, ...], ...]:
+        """Reverse the content of a road between the point 1 and 2 (included)."""
+        size = len(road)
+
+        if index1 < index2:
+            index_min = index1
+            index_max = index2
+        else:
+            index_min = index2
+            index_max = index1
+
+        begining = road[0:index_min]
+        middle = road[index_min:(index_max+1)]
+        middle.reverse()
+        ending = road[(index_max+1):size]
+
+        reversed_road = begining + middle + ending
+
+        return reversed_road
+
+    def __two_opt(self, road: list[int]) -> list[int]:
+        """
+        2-opt variance algorithm.
+
+        Algorithm which find the best mutation for this road by calculating
+        the distance for each mutation of road. The shortest distance will
+        determine the best muatation road.
+        """
+        best_distance = self._compute_road_fitness(road)
+        best_road = road
+
+        depot = self.__instance.get_depot()
+
+        road.remove(depot)  # remove the first depot
+        road.remove(depot)  # remove the last depot
+
+        for index_point1 in range(len(road)):
+            for index_point2 in range(len(road)):
+                if index_point1 != index_point2:
+                    tmp_road = self.__reversed_content(road, index_point1, index_point2)
+                    tmp_road.insert(0, depot)  # add the depot at the begining
+                    tmp_road.append(depot)  # add the depot at the end
+                    distance = self._compute_road_fitness(tmp_road)
+                    if distance < best_distance:
+                        best_distance = distance
+                        best_road = tmp_road
+
+        return best_road
+
+    def delete_empty_road(self, solution: list[list[int]]) -> list[list[int]]:
+        """Delete empty road which are represented by [0, 0] if depot is 0."""
+        depot = self.__instance.get_depot()
+        for road in solution:
+            if road == [depot, depot]:
+                solution.remove(road)
+        return solution
+
     def mutate(self) -> None:
-        """ Mutate the current individual according to its internal rules.
-            This changement is done in place.
         """
+        Mutate the current individual according to its internal rules.
+
+        This changement is done in place.
+        """
+        solution_tuple = self.get_roads()
+        solution = self.__tuple_to_list(solution_tuple)
+
+        choice = randrange(0, len(solution))
+        road = solution[choice]
+
+        road_without_chargers = self.__pull_off_chargers(road)
+
+        # algo 2-opt
+        mutant_road = self.__two_opt(road_without_chargers)
+
+        new_road = self._road_correction(mutant_road)
+        solution[choice] = new_road
+
+        solution = self.delete_empty_road(solution)
+
+        self._solution = self.__merge_roads(solution)
+
+    def __get_solutions(self, other: "ECVRPSolution") -> tuple[list[list[int]], list[list[int]]]:
```

Comment:
These functions are reducing the comprehension complexity issue, but not the duplication one.  Here is an example on how to effectively reduce the code duplication : 

```python
def __get_solutions(self) -> list[list[int]]:
        roads = self.__tuple_to_list(self.get_roads())

        # pull of charger for the crossover opration
        for index_road, road in enumerate(roads):
            roads[index_road] = self.__pull_off_chargers(road)

        return roads
```

and then : 

```python
s_1 = self.__get_solutions()
s_2 = other.__get_solutions()
```

This better follows the DRY principle. Note that the method would need to be renamed after this change.

Reply:
done